controllers/config.go

```diff
@@ -0,0 +1,146 @@
+	SecretControllerConfig *secret.Config
+}
+
+// InitFromFlags initializes the controller manager config from the provided CLI flag set.
+func InitFromFlags(fs *flag.FlagSet, cfg *ManagerConfig) {
+	flag.StringVar(&cfg.MetricsAddr, metricsAddrFlagName, defaultMetricsAddr, ""+
+		"The address the metric endpoint binds to.")
+	flag.BoolVar(&cfg.EnableLeaderElection, enableLeaderElectionFlagName, defaultEnableLeaderElection,
+		"Enable leader election for controller manager. Enabling this will ensure there is only one active controller manager.")
+	flag.StringVar(&cfg.LeaderElectionID, leaderElectionIDFlagName, defaultLeaderElectionID,
+		"Name of the resource that leader election will use for holding the leader lock")
+	flag.StringVar(&cfg.LeaderElectionResourceLock, leaderElectionResourceLockFlagName, defaultLeaderElectionResourceLock,
+		"Specifies which resource type to use for leader election. Supported options are 'endpoints', 'configmaps', 'leases', 'endpointsleases' and 'configmapsleases'.")
+	flag.BoolVar(&cfg.DisableLeaseCache, disableLeaseCacheFlagName, defaultDisableLeaseCache,
+		"Disable cache for lease.coordination.k8s.io resources.")
+	flag.BoolVar(&cfg.IgnoreOperationAnnotation, ignoreOperationAnnotationFlagName, defaultIgnoreOperationAnnotation,
+		"Specifies whether to ignore or honour the operation annotation on resources to be reconciled.")
+
+	cfg.EtcdControllerConfig = &etcd.Config{}
+	etcd.InitFromFlags(fs, cfg.EtcdControllerConfig)
+
+	cfg.CustodianControllerConfig = &custodian.Config{}
+	custodian.InitFromFlags(fs, cfg.CustodianControllerConfig)
+
+	cfg.CompactionControllerConfig = &compaction.Config{}
+	compaction.InitFromFlags(fs, cfg.CompactionControllerConfig)
+
+	cfg.EtcdCopyBackupsTaskControllerConfig = &etcdcopybackupstask.Config{}
+	etcdcopybackupstask.InitFromFlags(fs, cfg.EtcdCopyBackupsTaskControllerConfig)
+
+	cfg.SecretControllerConfig = &secret.Config{}
+	secret.InitFromFlags(fs, cfg.SecretControllerConfig)
+}
+
+// Validate validates the controller manager config.
+func (cfg *ManagerConfig) Validate() error {
+	if err := utils.ShouldBeOneOfAllowedValues("LeaderElectionResourceLock", getAllowedLeaderElectionResourceLocks(), cfg.LeaderElectionResourceLock); err != nil {
+		return err
+	}
+	if err := cfg.EtcdControllerConfig.Validate(); err != nil {
+		return err
+	}
+
+	if err := cfg.CustodianControllerConfig.Validate(); err != nil {
+		return err
+	}
+
+	if err := cfg.CompactionControllerConfig.Validate(); err != nil {
+		return err
+	}
+
+	if err := cfg.EtcdCopyBackupsTaskControllerConfig.Validate(); err != nil {
+		return err
+	}
+
+	if err := cfg.SecretControllerConfig.Validate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// getAllowedLeaderElectionResourceLocks gives the resource names that can be used for leader election.
+// NOTE: This should be changed as lease is the default choice now for leader election. There is no need
```

Comment:
`TODO`s make it easier to follow-up on technical debts.

```suggestion
// TODO: This should be changed as lease is the default choice now for leader election. There is no need
```

Reply:
done